File: proj/astree.cpp
#include <iomanip>
#include <iostream>
using namespace std;


#include "location.h"

#include "astree.h"
#include "util.h"
#include "attr.h"


ostream& operator<< (ostream& out, const ASTNode& tree) {
   out << get_parser_yytname (tree.symbol) << " \"" << 
          tree.lexinfo << "\" (" << tree.loc << " " << tree.syminfo ;

   if (tree.syminfo.attrbits[attr::CONST] || 
         tree.syminfo.attrbits[attr::VARIABLE] ||
         tree.syminfo.attrbits[attr::ASSIGN] ||
         tree.syminfo.attrbits[attr::BINOP] ||
         tree.syminfo.attrbits[attr::UNOP])
      out << " VR: "<< tree.vr ;
   out << " )";
   return out;
}

void ASTNode::print_tree (ostream& out, int depth) {
   for (int i = 0; i < depth; i++)
      out << "|  "; 
   out << *this << endl;
   for (astree_ptr child: children) child->print_tree (out, depth + 1);
}


ASTNode::~ASTNode() {
   for (astree_ptr child: children) erase (child);
}

astree_ptr ASTNode::adopt(astree_ptr child){
    children.push_back(child);
    return this;
}

astree_ptr ASTNode::adopt(initializer_list<astree_ptr> args){
    for (auto child : args){
        children.push_back(child);
    }
    return this;
}

astree_ptr ASTNode::adopt_as (int symbol_, astree_ptr& child) {
   return set_symbol(symbol_)->adopt(child);
}

astree_ptr ASTNode::adopt_as (int symbol_, initializer_list<astree_ptr> args) {
   return set_symbol(symbol_)->adopt(args);
}


astree_ptr ASTNode::set_symbol(int symbol_){
   symbol = symbol_;
   return this;
}

astree_ptr ASTNode::set_bits(attr attr_) {
   syminfo.attrbits[attr_] = 1;
   return this;
}

astree_ptr ASTNode::set_bits(initializer_list<attr> args){
   for (auto at : args){
      set_bits(at);
   }
   return this;
}

astree_ptr ASTNode::set_bits(astree_ptr& tree){
   syminfo.attrbits = tree->syminfo.attrbits;
   return this;
}

astree_ptr ASTNode::make (int symbol, const location& loc,
                         const string lexinfo) {
   return new ASTNode(symbol, loc, lexinfo);
}

astree_ptr ASTNode::mkopt(int symbol, initializer_list<astree_ptr> children) {
   location p = (*(children.begin()))-> loc;

   astree_ptr ret_ptr = make(symbol, p, "");
   ret_ptr->adopt(children);
   return ret_ptr;
}


void ASTNode::erase (astree_ptr& tree) {
   delete tree;
   tree = nullptr;
}

void ASTNode::erase (initializer_list<astree_ptr> children) {
   for (auto child : children){
      delete child;
      child = nullptr;
   }
}

File: rewrite/TestApplication/astree.cs
using System;
using System.Collections.Generic;
using location;
using attr;
using symnames;

namespace astree{
    public class ASTNode {
        int symbol;
        Location loc;
        Attributes syminfo;
        string lexinfo;

        // an array of children
        List<ASTNode> children;

        // associated vr
        string vr = "";

        public override string ToString(){
            string ret = Enum.GetName(typeof(SymType), symbol) + " \"" + lexinfo + "\" (" + loc.ToString() + " " + syminfo.ToString();
            if (syminfo[Attr.CONST] ||
                   syminfo[Attr.VARIABLE] ||
                   syminfo[Attr.ASSIGN] ||
                   syminfo[Attr.BINOP] ||
                   syminfo[Attr.UNOP])
            ret += " VR: " + vr;
            ret += " )";

            return ret;
        }

        public ASTNode(int symbol_ , Location loc_, string lexinfo_){
            symbol = symbol_;
            loc = loc_;
            lexinfo = lexinfo_;
            children = new List<ASTNode>();
        }
        public string PrintTree(int n=0){
            string ret = "";
            
            for (int i = 0; i < n; i++){
                ret += "    ";
            }

            ret += ToString()+ "\n";
            foreach (ASTNode c in children){
                ret += c.PrintTree(n+1);
            }
            return ret;
        }

        public bool IsLeafNode(){
            return syminfo[Attr.CONST] || syminfo[Attr.VARIABLE];
        }

        public bool IsBinopNode(){
            return syminfo[Attr.BINOP];
        }

        public bool IsUnopNode(){
            return syminfo[Attr.UNOP];
        }

        public bool IsCompNode(){
            return syminfo[Attr.COMP];
        }

        public bool isConvNode(){
            return syminfo[Attr.INT2FLOAT] || syminfo[Attr.FLOAT2INT];
        }
        
        public ASTNode adopt(ASTNode node){
            children.Add(node);
            return this;
        }

        public ASTNode adopt(List<ASTNode> args){
            foreach (ASTNode n in args){
                children.Add(n);
            }
            return this;
        }

        public ASTNode adopt_as (int symbol_, ASTNode child){
            return this.set_symbol(symbol_).adopt(child);
        }

        public ASTNode adopt_as (int symbol_, List<ASTNode> args){
            return this.set_symbol(symbol_).adopt(args);
        }

        public ASTNode set_symbol(int symbol_){
            symbol = symbol_;
            return this;

        }

        public ASTNode set_bits(Attr a){
            syminfo[a] = true;
            return this;
        }

        public ASTNode set_bits(List<Attr> args){ // I think use a list?
            foreach(Attr b in args){
                syminfo[b] = true;
            }
            return this;
        }

        public ASTNode set_bits(ASTNode node){
            syminfo.attrbits = node.syminfo.attrbits;
            return this;
        }

        static ASTNode make(int symbol_, Location loc_, string lexinfo){
            return new ASTNode(symbol_, loc_, lexinfo);
        }

        static ASTNode mkopt(int symbol, List<ASTNode> args){
            Location p = args[0].loc;

            ASTNode ret_ptr = ASTNode.make(symbol, p, "");
            ret_ptr.adopt(args);
            return ret_ptr;
        }
    }
}
